arrays_strings/rotation/rotation_solution.ipynb

```diff
@@ -109,15 +111,85 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## Unit Test"
+    "## Algorithm: Comparing In Place\n",
+    "\n",
+    "Notes:\n",
+    "* This is taking the same high-level approach as the previous solution, but it is operating on the strings in place without creating any new string.\n",
+    "* This is a bit more complex and less readable.\n",
+    "* This approach is not using a built-in Python collection operator, so it will be slower.\n",
+    "\n",
+    "Complexity:\n",
+    "* Time: O(n)\n",
+    "* Space: O(1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Code: Comparing In Place"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": 2,
-   "metadata": {
-    "collapsed": false
-   },
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class RotationInPlace(object):\n",
+    "\n",
+    "    def is_substring(self, s1, s2):\n",
+    "        # We already assured the strings are the same length; this is for readability\n",
+    "        length = len(s1)\n",
+    "\n",
+    "        # If both strings are empty, they are trivially a rotation of one another\n",
+    "        if length == 0:\n",
+    "            return True\n",
+    "\n",
+    "        i1, i2, matching_count = 0, 0, 0\n",
+    "        # This loop achieves the same as concatenating a string with itself\n",
+    "        for _ in range(2*length):\n",
+    "            \n",
+    "            # Count consecutive character matches.\n",
+    "            if s1[i1] == s2[i2]:\n",
+    "                matching_count += 1\n",
+    "                \n",
+    "                # If the count is equal to the length of the strings, we have found a rotation\n",
+    "                if matching_count == length:\n",
+    "                    return True\n",
+    "                \n",
+    "                i2 += 1\n",
+    "            # If there is a mismatch, start the count over again\n",
+    "            else:\n",
+    "                matching_count, i2 = 0, 0\n",
+    "\n",
+    "            # Increment i1 and wrap around if necessary\n",
+    "            i1 += 1\n",
```

Comment:
```python
i1 += 1
if i1 == length:
    i1 = 0
```

can be replaced by:

```python
i1 = (i1 + 1) % length
```

which is a common idiom for wrap-around semantics in iteration.